pkg/agent/billing/queue.go

```diff
@@ -0,0 +1,79 @@
+package billing
+
+// Implementation of the event queue for mediating event generation and event sending.
+//
+// The "public" (ish - it's all one package) types are eventQueuePuller and eventQueuePusher, two
+// halves of the same queue. Each half is only safe for use from a single thread, but *together*
+// they can be used in separate threads.
+
+import (
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/exp/slices"
+
+	"github.com/neondatabase/autoscaling/pkg/util"
+)
+
+// this is generic just so there's less typing - "billing.IncrementalEvent" is long!
+type eventQueueInternals[E any] struct {
+	mu        sync.Mutex
+	items     []E
+	sizeGauge prometheus.Gauge
+}
+
+type eventQueuePuller[E any] struct {
+	internals *eventQueueInternals[E]
+}
+
+type eventQueuePusher[E any] struct {
+	internals *eventQueueInternals[E]
+}
+
+func newEventQueue[E any](sizeGauge prometheus.Gauge) (eventQueuePusher[E], eventQueuePuller[E]) {
+	internals := &eventQueueInternals[E]{
+		mu:        sync.Mutex{},
+		items:     nil,
+		sizeGauge: sizeGauge,
+	}
+	return eventQueuePusher[E]{internals}, eventQueuePuller[E]{internals}
+}
+
+// NB: must hold mu
+func (qi *eventQueueInternals[E]) updateGauge() {
+	qi.sizeGauge.Set(float64(len(qi.items)))
+}
```

Comment:
this probably doesn't need to be a pointer method. 

also prom gauges are threadsafe, so I think you can make the locking story more hygenic, by passing the size as an argument rather.

Reply:
> this probably doesn't need to be a pointer method.

The linter gets cranky otherwise because `eventQueueInternals` contains a mutex.

> I think you can make the locking story more hygenic, by passing the size as an argument rather.

I'm not sure that's better - doing it that way, I'd need to pass in `len(qi.items)` both times `updateGauge` is called, which... idk, feels less encapsulated.